framework/db/BaseActiveRecord.php

```diff
@@ -606,15 +606,19 @@ public function markAttributeDirty($name)
      */
     public function isAttributeChanged($name, $identical = true)
     {
-        if (isset($this->_attributes[$name], $this->_oldAttributes[$name])) {
-            if ($identical) {
-                return $this->_attributes[$name] !== $this->_oldAttributes[$name];
-            }
+        if (!array_key_exists($name, $this->_attributes)) {
+            return false;
+        }
+
+        if (!array_key_exists($name, $this->_oldAttributes)) {
+            return true;
+        }
 
-            return $this->_attributes[$name] != $this->_oldAttributes[$name];
+        if ($identical) {
+            return $this->_attributes[$name] !== $this->_oldAttributes[$name];
         }
 
-        return isset($this->_attributes[$name]) || isset($this->_oldAttributes[$name]);
```

Comment:
You could just change the 2nd `isset` in this line to use `array_key_exists` and it would work?

Reply:
Because an attribute can have any correct values like ``null, 0, 1, 55, 'string' or 'whatever'`` there is no reason to keep any test with ``isset`` that would introduce a false special case for ``null``, and then rely on magical coincidence to treat this special case.
The proposed code is pretty straight forward, works with nulls and saves a test level.

Reply:
It is actually introducing faulty behavior as far as I can tell... See my comment in the test.

Reply:
This test is already done line 1839. But I guess we disagree on the assertion line 1840.

IMHO when an attribute is unset, the model is not dirty (the state of the attribute is just unknown, not modified).
If you complete your code with ``$order->save()``, you will see that is doesn't do anything because there is no dirty attributes.

So it is more consistent to have ``$order->isAttributeChanged('total')`` returning ``false`` when ``$order->getDirtyAttributes()`` is not returning the attribute neither.

Reply:
I saw that line, but it is part of a bigger test, where you're unsetting it and confirming that it was reset; so just to be sure I wanted the singular test that tests very specifically the behavior I want.

In my opinion of if the record the database has value `10` for the `total` field we should have this behavior (and I was under the impression that this is the behavior of the framework):
```php
$order = Order::findOne(1);
$this->assertEquals(10, $order->total);
unset($order->total);
$this->assertTrue($order->isAttributeChanged('total'));
$order->save(); // This should save `null` in the `total` field.
$order = Order::findOne(1);
$this->assertNull($order->total);
$this->total = null;
$this->assertFalse($this->isAttributeChanged('total'));
```

Note that if that's not the behavior Yii currently has then your tests / code makes sense (and I have to do some code checking to see if there are no bugs).

Reply:
````php
$order->save(); // This should save `null` in the `total` field.
//Actually no, it doesn't save anything
````
The current behavior is not saving anything. And it makes sense to me.
An unset value is not equal to null nor 10. Unset means you just don't know what's in it, so you just don't change any existing value.

Reply:
Okay, that explains my confusion. If the behavior is as you describe then I'm fine with the code changes and indeed do not see a change in behavior.
Thanks for being patient and providing clear arguments in the discussion! :1st_place_medal:

Reply:
> Unset means you just don't know what's in it, so you just don't change any existing value.

That's debatable. IMO unset means that attribute has no value. Update query should reflect this change, and set column to `null` (which is representation of "no value").

IMO this is bug in `__unset()`. Current behavior is not even consistent with method description:


https://github.com/yiisoft/yii2/blob/c1574f3cb7024d4833604ee41eb7066e280653f1/framework/db/BaseActiveRecord.php#L346

Reply:
But making it consistent would be a big BC break, and it would also fully break partial updates in AR; since retrieving only specific fields via `->select()` will not set those attributes. If we then call save we don't want all unsaved attributes to be saved as `null`.
So for 2.x this PR is good and if we want changed behavior a new discussion for 3.0 is in order, right @rob006?
In that case maybe we should move the discussion and let this PR go through?

Reply:
@SamMousa I'm talking only about `__unset()` behavior. If you change `unset($this->_attributes[$name]);` to `$this->_attributes[$name] = null;`, then behavior for partial select should not be affected at all.

Reply:
But that's not consistent with `isAttributeChanged` then right? 
Basically if `isAttributeChanged` returns `true` we should expect that same attribute to be sent to the database when calling save.
Currently according to @CedricYii this is not the case when we use `unset($record->attribute)`; in which case his implementation of the `isAttributeChanged` function makes sense.

If we change behavior of `__unset()`, then we are breaking backwards compatibility. Also this PR has nothing to do with that change, other than the fact that the `isAttributeChanged` might need to be updated again if we change behavior of `__unset()`.

Reply:
> But that's not consistent with isAttributeChanged then right?

This is not consistent with **new implementation** of `isAttributeChanged()` (for old one there was no difference between `null` value and missing key), so this PR makes this bug more visible. I suggest to fix it before we change how `isAttributeChanged()` work and introduce this faulty behavior also for `isAttributeChanged()`.

Reply:
But is it worth it to fix this and break BC in 2.x?
Since we've had no issues pop up related to this bug, I think it might not be worth it...
We could also leave the code as is, don't fix the bug and change behavior and remove the bug in 3.x...

Reply:
Maybe I'm missing something but:
1. Currently `isAttributeChanged()`  and `getDirtyAttributes()` is inconsistent for `unset($model->attribute)` case.
2. This PR changes `isAttributeChanged()` behavior to be consistent with `getDirtyAttributes()` (BC break for `isAttributeChanged()`). But IMO current `getDirtyAttributes()` behavior is incorrect, so we're spreading incorrect behavior to other places to.
3. I propose to change `getDirtyAttributes()` to be consistent with result of old `isAttributeChanged()` implementation (BC break for `getDirtyAttributes()`).

One way or another we will break BC in some way.

Reply:
> Currently unset() is not setting attribute to null (and has never done it).

Yes, it does: https://github.com/yiisoft/yii2/pull/16709#issuecomment-420936871

You're basically proposing situation where `$model->getAttribute('attribute') != $model->getOldAttribute('attribute')` but `$model->isAttributeChanged('attribute')` will return `false`. 

I'm feeling that this is going to wrong direction...